File: ios/KovanicaWallet/ViewModels/WalletViewModel.swift
import Foundation

@MainActor
final class WalletViewModel: ObservableObject {
    enum WalletUIState: Equatable {
        case loading
        case error(String)
        case data(WalletData)
    }

    struct WalletData: Equatable {
        var address: String = ""
        var balance: Int64 = 0
        var balanceFormatted: String = "0.00"
        var network: String = ""
        var tokenSymbol: String = NetworkConstants.TOKEN_SYMBOL
        var tokenName: String = NetworkConstants.TOKEN_NAME
        var blockHeight: Int = 0
        var minFee: Int64 = 0
        var faucetAvailable: Bool = false
        var history: [HistoryEntry] = []
        var historyOffset: Int = 0
        var historyHasMore: Bool = true
        var isLoadingHistory: Bool = false
        var faucetLoading: Bool = false
        var faucetMessage: String?
        var nodeURL: String
    }

    @Published var state: WalletUIState = .loading
    @Published var nodeURL: String = NetworkConstants.KOVANICA_API_DEFAULT

    private let defaultsKey = "kvnc_node_url"
    private var repository: WalletRepository

    init() {
        let saved = UserDefaults.standard.string(forKey: "kvnc_node_url") ?? NetworkConstants.KOVANICA_API_DEFAULT
        self.nodeURL = saved
        self.repository = WalletRepository(baseURL: saved)
    }

    var data: WalletData? {
        if case .data(let d) = state { return d }
        return nil
    }

    func load() async {
        state = .loading
        do {
            let head = try await repository.fetchHead()
            let stateResp = try await repository.fetchState()

            var wd = WalletData()
            wd.network = head.network
            wd.blockHeight = head.blocks
            wd.minFee = head.minFee
            wd.faucetAvailable = stateResp.faucet
            wd.nodeURL = nodeURL

            if let addr = try? await repository.fetchBalance(address: "kvnc_test_address_placeholder") {
                wd.address = addr.address
                wd.balance = addr.balance
                wd.balanceFormatted = TokenFormatter.formatBalanceDisplay(addr.balance)
            }

            let history = try await repository.fetchHistory(address: wd.address, limit: 20, offset: 0)
            wd.history = history
            wd.historyOffset = history.count
            wd.historyHasMore = history.count >= 20

            state = .data(wd)
        } catch {
            state = .error(error.localizedDescription)
        }
    }

    func loadWalletAddress(_ address: String) async {
        state = .loading
        do {
            let head = try await repository.fetchHead()
            let stateResp = try await repository.fetchState()
            let balanceResp = try await repository.fetchBalance(address: address)
            let history = try await repository.fetchHistory(address: address, limit: 20, offset: 0)

            var wd = WalletData()
            wd.address = balanceResp.address
            wd.balance = balanceResp.balance
            wd.balanceFormatted = TokenFormatter.formatBalanceDisplay(balanceResp.balance)
            wd.network = head.network
            wd.blockHeight = head.blocks
            wd.minFee = head.minFee
            wd.faucetAvailable = stateResp.faucet
            wd.nodeURL = nodeURL
            wd.history = history
            wd.historyOffset = history.count
            wd.historyHasMore = history.count >= 20

            state = .data(wd)
        } catch {
            state = .error(error.localizedDescription)
        }
    }

    func requestFaucet() async {
        guard var wd = data, !wd.faucetLoading else { return }
        wd.faucetLoading = true
        wd.faucetMessage = nil
        state = .data(wd)

        do {
            let resp = try await repository.requestFaucet(address: wd.address)
            if let msg = resp.error {
                var updated = wd
                updated.faucetLoading = false
                updated.faucetMessage = msg
                state = .data(updated)
            } else if resp.ok == true || resp.tx != nil {
                try? await Task.sleep(nanoseconds: 2_000_000_000)
                let balanceResp = try await repository.fetchBalance(address: wd.address)
                var updated = wd
                updated.balance = balanceResp.balance
                updated.balanceFormatted = TokenFormatter.formatBalanceDisplay(balanceResp.balance)
                updated.faucetLoading = false
                updated.faucetMessage = "1 KVNC received"
                let history = try await repository.fetchHistory(address: wd.address, limit: 20, offset: 0)
                updated.history = history
                updated.historyOffset = history.count
                updated.historyHasMore = history.count >= 20
                state = .data(updated)
            } else {
                var updated = wd
                updated.faucetLoading = false
                updated.faucetMessage = "Faucet request sent"
                state = .data(updated)
            }
        } catch {
            var updated = wd
            updated.faucetLoading = false
            updated.faucetMessage = "Error: \(error.localizedDescription)"
            state = .data(updated)
        }
    }

    func loadMoreHistory() async {
        guard var wd = data, wd.historyHasMore, !wd.isLoadingHistory else { return }
        wd.isLoadingHistory = true
        state = .data(wd)

        do {
            let more = try await repository.fetchHistory(address: wd.address, limit: 20, offset: wd.historyOffset)
            var updated = wd
            updated.history.append(contentsOf: more)
            updated.historyOffset += more.count
            updated.historyHasMore = more.count >= 20
            updated.isLoadingHistory = false
            state = .data(updated)
        } catch {
            var updated = wd
            updated.isLoadingHistory = false
            state = .data(updated)
        }
    }

    func setNodeURL(_ url: String) {
        nodeURL = url
        UserDefaults.standard.set(url, forKey: defaultsKey)
        repository = WalletRepository(baseURL: url)
    }

    func refresh() async {
        guard let wd = data else {
            await load()
            return
        }
        await loadWalletAddress(wd.address)
    }
}

File: android/app/src/main/java/com/kovanica/wallet/ui/WalletViewModel.kt
package com.kovanica.wallet.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.kovanica.wallet.KovanicaWalletApp
import com.kovanica.wallet.data.WalletRepository
import com.kovanica.wallet.model.AddressInfo
import com.kovanica.wallet.model.ATOM
import com.kovanica.wallet.model.BootstrapInfo
import com.kovanica.wallet.model.FeeEstimate
import com.kovanica.wallet.model.HeadInfo
import com.kovanica.wallet.model.HistoryEntry
import com.kovanica.wallet.model.KVNC_SYMBOL
import com.kovanica.wallet.model.KVNC_TOKEN_NAME
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.math.BigDecimal
import java.math.RoundingMode

data class WalletUiState(
    val isLoading: Boolean = false,
    val isRefreshing: Boolean = false,
    val error: String? = null,
    val address: String = "",
    val balanceAtoms: Long = 0L,
    val head: HeadInfo? = null,
    val bootstrap: BootstrapInfo? = null,
    val feeEstimate: FeeEstimate? = null,
    val history: List<HistoryEntry> = emptyList(),
    val historyHasMore: Boolean = false,
    val faucetBusy: Boolean = false,
    val faucetMessage: String? = null,
) {
    val balanceFormatted: String
        get() {
            val kvnc = BigDecimal(balanceAtoms)
                .divide(BigDecimal(ATOM.toString()), 8, RoundingMode.HALF_UP)
                .stripTrailingZeros()
                .toPlainString()
            return "$kvnc KVNC"
        }

    val tokenName: String get() = KVNC_TOKEN_NAME
    val tokenSymbol: String get() = KVNC_SYMBOL
    val network: String get() = head?.network ?: bootstrap?.network ?: ""
}

class WalletViewModel(
    private val app: KovanicaWalletApp,
) : ViewModel() {

    private var repository: WalletRepository = app.repository()

    val nodeUrl: String get() = app.nodeUrl

    private val _uiState = MutableStateFlow(WalletUiState(address = repository.address))
    val uiState: StateFlow<WalletUiState> = _uiState

    private var historyOffset = 0
    private val historyPageSize = 25

    init {
        load()
    }

    fun setAddress(address: String) {
        repository.setAddress(address)
        _uiState.value = _uiState.value.copy(address = address)
        load()
    }

    fun setNodeUrl(url: String) {
        app.nodeUrl = url
        repository = app.repository()
        load()
    }

    fun load() {
        val addr = repository.address
        if (addr.isBlank()) {
            _uiState.value = _uiState.value.copy(isLoading = false, error = "No watch address set")
            return
        }
        _uiState.value = _uiState.value.copy(isLoading = true, error = null)
        viewModelScope.launch {
            runCatching {
                val head = repository.loadHead()
                val bootstrap = repository.loadBootstrap()
                val fee = repository.loadFeeEstimate()
                val addrInfo = repository.loadAddress()
                val firstPage = repository.loadHistory(limit = historyPageSize, offset = 0)
                historyOffset = firstPage.size
                Trip(head, bootstrap, fee, addrInfo, firstPage)
            }.fold(
                onSuccess = { t ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = null,
                        address = addr,
                        head = t.head,
                        bootstrap = t.bootstrap,
                        feeEstimate = t.fee,
                        balanceAtoms = t.addr.balance,
                        history = t.history,
                        historyHasMore = t.history.size >= historyPageSize,
                    )
                },
                onFailure = { e ->
                    _uiState.value = _uiState.value.copy(
                        isLoading = false,
                        error = e.message ?: "Load failed",
                    )
                },
            )
        }
    }

    fun loadMoreHistory() {
        viewModelScope.launch {
            runCatching {
                repository.loadHistory(limit = historyPageSize, offset = historyOffset)
            }.onSuccess { page ->
                historyOffset += page.size
                val prev = _uiState.value
                _uiState.value = prev.copy(
                    history = prev.history + page,
                    historyHasMore = page.size >= historyPageSize,
                    isRefreshing = false,
                )
            }
        }
    }

    fun refresh() {
        _uiState.value = _uiState.value.copy(isRefreshing = true)
        load()
    }

    fun requestFaucet() {
        if (_uiState.value.faucetBusy || repository.address.isBlank()) return
        _uiState.value = _uiState.value.copy(faucetBusy = true, faucetMessage = null)
        viewModelScope.launch {
            runCatching { repository.requestFaucet() }.fold(
                onSuccess = {
                    _uiState.value = _uiState.value.copy(
                        faucetBusy = false,
                        faucetMessage = "Faucet request sent",
                    )
                    load()
                },
                onFailure = { e ->
                    _uiState.value = _uiState.value.copy(
                        faucetBusy = false,
                        faucetMessage = e.message ?: "Faucet request failed",
                    )
                },
            )
        }
    }

    override fun onCleared() {
        super.onCleared()
    }

    private data class Trip(
        val head: HeadInfo,
        val bootstrap: BootstrapInfo,
        val fee: FeeEstimate,
        val addr: AddressInfo,
        val history: List<HistoryEntry>,
    )

    companion object {
        val Factory: ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val app = this[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY] as KovanicaWalletApp
                WalletViewModel(app)
            }
        }
    }
}
